### ISSUE TITLE
[Question] page.waitForNavigation not working as expected for previously visited URL

### ISSUE THREAD

**Issue by soulchainer** [microsoft/playwright#6078]
`playwright` version: 1.10.0

Sorry if I make mistakes, I'm just evaluating Playwright to write tests with it.

I have this test, where I'm checking if:
1. a page navigates to the proper card details page when you click in a card from the list of cards
2. the link to navigate back from the card details page to the list of card also works fine

But I don't see consistency in the way the test is passing or failing. Check the comments in the code.

```js
import { chromium } from 'playwright'
import { getDocument, queries, waitFor } from 'playwright-testing-library'

const BASE_URL = 'http://localhost:3000/'

let browser
let page

beforeAll(async () => { browser = await chromium.launch() })
afterAll(async () => { await browser.close() })
beforeEach(async () => {
  page = await browser.newPage()
  await page.goto(BASE_URL);
})
afterEach(async () => { await page.close() })

describe('home page', () => {
  // other test cases
  it('will take the user to a character detail page when its card is clicked, and back', async () => {
    const { getAllByTestId, getByLabelText } = queries
    const $document = await getDocument(page)

    await (await getAllByTestId($document, 'card'))[5].click()
    await page.waitForNavigation({ url: `${BASE_URL}people/6` })
    await (await getByLabelText($document, 'back to main list')).click()
    // the test breaks at this line, it's like it can't detect the page navigating to BASE_URL (triggers an error in the `getAllByTestId` line, clicking in the card, or a test timeout error, waiting for the navigation to happen)
    await page.waitForNavigation({ url: BASE_URL })
    // if we replace the previous line with this one, the test passes, because the navigation, obviously, happened (and fast)
    expect(await page.url()).toBe(BASE_URL)
  })
})
```

This seems inconsistent/counterintuitive to me. IDK why on first navigation, the `page.waitForNavigation` works and then, when I'm about to (re)visit the previous page, then it doesn't work and I need to check the navigation there using this other way, instead of it being consistent and being able to detect the navigation in the same, consistent way.

In both cases navigation is done using the `<Link>` component of Next.js.

`<Link href="/people/[id]" as={path} prefetch={false}>...</Link>`
`<Link href="/"><a aria-label="back to main list"></a></Link>`

So it's not like they're doing anything different at all.

As I said, this works this way, but I would appreciate any comment about why navigation has to be tested differently in one case and the other.

**Comment by pavelfeldman** [comment]
https://playwright.dev/docs/navigations#asynchronous-navigation would be a good reference. waitForNavigation is waiting for the subsequent navigation. If navigation is already committed by the time you call waitForNavigation, it'll stall waiting for the next one. Promise.all from the snippet takes care of that.

Playwright 1.11 will contain `page.waitForURL` that'll ensure url matches and if not, waits for subsequent navigation to that url pattern. That new method won't require Promise.all, but will require URL param to be there.

**Comment by soulchainer** [comment]
Ufff, silly me :blush: .

The thing is that I already read docs, but not these: [the `Page` class](https://playwright.dev/docs/api/class-page#pagewaitfornavigationoptions) ones, after reading #2142. And in there they also used `Promise.all` in the example, but I just didn't take it like «this must be used this way» but as «this could be used this way, for convenience», so I did a quick read and didn't pay much attention then to the code comments in the example.

In the docs you linked it's so clear it should be used this way :).
I now understand that the part that was working as I expected was doing so only by fluke, and it could eventually fail because of a race condition, so I changed both and now it works like a charm :). And keeping consistency :).

Thanks for your patience and help! :100: .
It's good to know about `page.waitForURL`. I have been using the `url` param always till now, so it will come in handy when it gets released.

I'm closing this then :). Thanks again for your help.
